File: packages/sdk/src/sessions.ts
import { DaemonHttpError } from './client.js';
import type { DaemonClient } from './client.js';

export type SessionKind = 'chat' | 'telegram' | 'checkin' | 'job' | 'helper';
export type SessionOrigin =
  | 'web'
  | 'api'
  | 'telegram'
  | 'schedule'
  | 'job'
  | 'delegation'
  | 'peer';
export type SessionTitleSource =
  | 'first_message'
  | 'generated'
  | 'owner'
  | 'system';

export interface SessionCapabilities {
  send: boolean;
  steer: boolean;
  stop: boolean;
  rename: boolean;
  archive: boolean;
  delete: boolean;
  compact: boolean;
  export: boolean;
}

export interface SessionSummary {
  text: string;
  throughMessageId: string;
  createdAtMs: number;
  sourceMessageCount: number;
}

export interface SessionContextTrimmed {
  droppedThroughMessageId: string;
  atMs: number;
}

/** Why a search result matched; `messageId` is null for a title match. */
export interface SessionMatch {
  messageId: string | null;
  snippet: string;
}

/** A session with the daemon's derived fields (spec §3.2). */
export interface Session {
  id: string;
  agentId: string;
  /** The transcript room; equals `id` except for mapped legacy rooms. */
  roomId: string;
  kind: SessionKind;
  origin: SessionOrigin;
  title: string;
  titleSource: SessionTitleSource;
  createdAtMs: number;
  lastActivityAtMs: number;
  lastReadAtMs: number | null;
  archived: boolean;
  parentSessionId: string | null;
  parentRunId: string | null;
  parentAgentId: string | null;
  summary: SessionSummary | null;
  contextTrimmed: SessionContextTrimmed | null;
  messageCount: number;
  preview: string | null;
  activeRuns: number;
  pendingApprovals: number;
  unread: boolean;
  capabilities: SessionCapabilities;
  match?: SessionMatch;
}

export interface SessionPage {
  sessions: Session[];
  nextCursor: string | null;
}

export interface SessionListOptions {
  kind?: SessionKind;
  /** true lists only archived sessions; the daemon default is false. */
  archived?: boolean;
  q?: string;
  cursor?: string;
  /** 1–200; the daemon default is 50. */
  limit?: number;
  /** The daemon default is true. */
  includeHelpers?: boolean;
  signal?: AbortSignal;
}

export interface SessionMessageAttachment {
  type: 'file' | 'image' | 'url';
  name: string;
}

export interface SessionMessage {
  id: string;
  role: 'user' | 'assistant' | 'system' | 'tool';
  text: string;
  attachments: SessionMessageAttachment[];
  metadata: Record<string, unknown>;
  createdAtMs: number;
  /** Present on silent check-in messages when `includeHidden` is set. */
  hidden?: boolean;
}

export interface SessionMessagePage {
  /** Oldest to newest within the page. */
  messages: SessionMessage[];
  /**
   * The `before` cursor for the next (older) page, or null once history is
   * exhausted. A page degraded by an unreadable history store returns a
   * cursor here instead of null even when the hot tail itself is exhausted;
   * requesting that cursor then fails with a 503 rather than the page
   * quietly claiming there is nothing more.
   */
  nextBefore: string | null;
}

export interface SessionMessageOptions {
  before?: string;
  limit?: number;
  includeHidden?: boolean;
  signal?: AbortSignal;
}

export interface SessionUpdateInput {
  title?: string;
  archived?: boolean;
  lastReadAtMs?: number;
}

/**
 * Thrown only by `SessionsClient.list`. The daemon answers an unknown route
 * with the same JSON 404 as a real not-found, so a 404 from `GET
 * /api/agents/{id}/sessions` alone cannot tell "the daemon predates the
 * sessions routes" apart from "the agent does not exist" (Controller ruling
 * 2, M2 pre-flight audit; spec §13.4). `list` resolves the ambiguity by
 * probing `GET /api/agents/{id}`: if the agent exists, the sessions route is
 * the thing missing, so it raises this instead of the plain `DaemonHttpError`.
 * `code` is a stable string a caller can check without importing this class.
 */
export class DaemonTooOldError extends Error {
  readonly code = 'daemon_too_old' as const;

  constructor(readonly agentId: string) {
    super(
      `The daemon does not support sessions for agent "${agentId}" yet; update the daemon.`,
    );
    this.name = 'DaemonTooOldError';
  }
}

export class SessionsClient {
  constructor(private readonly client: DaemonClient) {}

  async list(
    agentId: string,
    options: SessionListOptions = {},
  ): Promise<SessionPage> {
    const search = new URLSearchParams();
    if (options.kind) search.set('kind', options.kind);
    if (options.archived !== undefined)
      search.set('archived', String(options.archived));
    if (options.q) search.set('q', options.q);
    if (options.cursor) search.set('cursor', options.cursor);
    if (options.limit !== undefined) search.set('limit', String(options.limit));
    if (options.includeHelpers !== undefined)
      search.set('includeHelpers', String(options.includeHelpers));
    try {
      return await this.client.requestJson<SessionPage>(
        withQuery(sessionsPath(agentId), search),
        { signal: options.signal },
      );
    } catch (error) {
      if (error instanceof DaemonHttpError && error.status === 404) {
        if (await this.agentExists(agentId)) {
          throw new DaemonTooOldError(agentId);
        }
      }
      throw error;
    }
  }

  /** Whether `GET /api/agents/{agentId}` answers 200 (Ruling 2, M2 pre-flight audit). */
  private async agentExists(agentId: string): Promise<boolean> {
    try {
      await this.client.agents.get(agentId);
      return true;
    } catch {
      return false;
    }
  }

  async get(
    agentId: string,
    sessionId: string,
    options: { signal?: AbortSignal } = {},
  ): Promise<Session> {
    const response = await this.client.requestJson<{ session: Session }>(
      sessionPath(agentId, sessionId),
      { signal: options.signal },
    );
    return response.session;
  }

  async messages(
    agentId: string,
    sessionId: string,
    options: SessionMessageOptions = {},
  ): Promise<SessionMessagePage> {
    const search = new URLSearchParams();
    if (options.before) search.set('before', options.before);
    if (options.limit !== undefined) search.set('limit', String(options.limit));
    if (options.includeHidden !== undefined)
      search.set('includeHidden', String(options.includeHidden));
    return this.client.requestJson<SessionMessagePage>(
      withQuery(`${sessionPath(agentId, sessionId)}/messages`, search),
      { signal: options.signal },
    );
  }

  async create(
    agentId: string,
    input: { title?: string } = {},
  ): Promise<Session> {
    const response = await this.client.requestJson<{ session: Session }>(
      sessionsPath(agentId),
      { method: 'POST', body: input },
    );
    return response.session;
  }

  async update(
    agentId: string,
    sessionId: string,
    patch: SessionUpdateInput,
  ): Promise<Session> {
    const response = await this.client.requestJson<{ session: Session }>(
      sessionPath(agentId, sessionId),
      { method: 'PATCH', body: patch },
    );
    return response.session;
  }

  async remove(agentId: string, sessionId: string): Promise<void> {
    await this.client.requestJson<{ deleted: boolean }>(
      sessionPath(agentId, sessionId),
      { method: 'DELETE' },
    );
  }

  /** The full transcript as Markdown, including archived history and
   *  messages kept only in the history store, with silent check-in turns
   *  marked rather than hidden. */
  async exportMarkdown(
    agentId: string,
    sessionId: string,
    options: { signal?: AbortSignal } = {},
  ): Promise<string> {
    return this.client.requestText(
      `${sessionPath(agentId, sessionId)}/export`,
      {
        signal: options.signal,
      },
    );
  }
}

function sessionsPath(agentId: string): string {
  return `/api/agents/${encodeURIComponent(agentId)}/sessions`;
}

function sessionPath(agentId: string, sessionId: string): string {
  return `${sessionsPath(agentId)}/${encodeURIComponent(sessionId)}`;
}

function withQuery(path: string, search: URLSearchParams): string {
  const query = search.toString();
  return query ? `${path}?${query}` : path;
}

File: hosts/rust-daemon/src/routes/contracts/sessions.rs
//! Session route bodies (spec §3.3).

use std::collections::BTreeMap;

use anima_core::{AttachmentType, MessageRole};
use serde::{Deserialize, Serialize};
use serde_json::Value;
use utoipa::ToSchema;

use super::shared::data_value_to_json;
use crate::sessions::views::{MessagePage, PageMessage, SessionPage, SessionView};
use crate::sessions::SessionCapabilities;

/// Message metadata the session routes expose: spec §3.3's list plus `kind`
/// (check-in prompts) and `source` (Telegram turns), and what a historical
/// tool card or step label shows: a tool result's `toolStatus` and
/// `toolDurationMs`, and `incomplete` on a model call's unfinished text.
/// `taskResult` stays hidden: it repeats the whole tool result.
pub(crate) const EXPOSED_MESSAGE_METADATA: [&str; 15] = [
    "toolCalls",
    "toolCallId",
    "stepId",
    "runId",
    "toolStatus",
    "toolDurationMs",
    "incomplete",
    "stopped",
    "revised",
    "steer",
    "skill",
    "clientRequestId",
    "communication",
    "kind",
    "source",
];

#[derive(Clone, Debug, Serialize, ToSchema)]
#[serde(rename_all = "camelCase")]
pub(crate) struct SessionCapabilitiesResponse {
    pub(crate) send: bool,
    pub(crate) steer: bool,
    pub(crate) stop: bool,
    pub(crate) rename: bool,
    pub(crate) archive: bool,
    pub(crate) delete: bool,
    pub(crate) compact: bool,
    pub(crate) export: bool,
}

impl From<SessionCapabilities> for SessionCapabilitiesResponse {
    fn from(value: SessionCapabilities) -> Self {
        Self {
            send: value.send,
            steer: value.steer,
            stop: value.stop,
            rename: value.rename,
            archive: value.archive,
            delete: value.delete,
            compact: value.compact,
            export: value.export,
        }
    }
}

#[derive(Clone, Debug, Serialize, ToSchema)]
#[serde(rename_all = "camelCase")]
pub(crate) struct SessionSummaryResponse {
    pub(crate) text: String,
    pub(crate) through_message_id: String,
    pub(crate) created_at_ms: u64,
    pub(crate) source_message_count: usize,
}

#[derive(Clone, Debug, Serialize, ToSchema)]
#[serde(rename_all = "camelCase")]
pub(crate) struct SessionContextTrimmedResponse {
    pub(crate) dropped_through_message_id: String,
    pub(crate) at_ms: u64,
}

#[derive(Clone, Debug, Serialize, ToSchema)]
#[serde(rename_all = "camelCase")]
pub(crate) struct SessionMatchResponse {
    /// `null` when only the title matched.
    pub(crate) message_id: Option<String>,
    pub(crate) snippet: String,
}

#[derive(Clone, Debug, Serialize, ToSchema)]
#[serde(rename_all = "camelCase")]
pub(crate) struct SessionResponse {
    pub(crate) id: String,
    pub(crate) agent_id: String,
    /// The transcript room: the id itself except for mapped legacy rooms.
    pub(crate) room_id: String,
    pub(crate) kind: String,
    pub(crate) origin: String,
    pub(crate) title: String,
    pub(crate) title_source: String,
    pub(crate) created_at_ms: u64,
    pub(crate) last_activity_at_ms: u64,
    pub(crate) last_read_at_ms: Option<u64>,
    pub(crate) archived: bool,
    pub(crate) parent_session_id: Option<String>,
    pub(crate) parent_run_id: Option<String>,
    pub(crate) parent_agent_id: Option<String>,
    pub(crate) summary: Option<SessionSummaryResponse>,
    pub(crate) context_trimmed: Option<SessionContextTrimmedResponse>,
    pub(crate) message_count: usize,
    pub(crate) preview: Option<String>,
    pub(crate) active_runs: usize,
    /// Always 0 until approvals exist (M4).
    pub(crate) pending_approvals: usize,
    pub(crate) unread: bool,
    pub(crate) capabilities: SessionCapabilitiesResponse,
    #[serde(rename = "match", skip_serializing_if = "Option::is_none")]
    pub(crate) matched: Option<SessionMatchResponse>,
}

impl From<&SessionView> for SessionResponse {
    fn from(view: &SessionView) -> Self {
        let record = &view.record;
        Self {
            id: record.id.clone(),
            agent_id: record.agent_id.clone(),
            room_id: record.room_id().to_string(),
            kind: record.kind.as_str().into(),
            origin: record.origin.as_str().into(),
            title: record.title.clone(),
            title_source: record.title_source.as_str().into(),
            created_at_ms: record.created_at_ms,
            last_activity_at_ms: record.last_activity_at_ms,
            last_read_at_ms: record.last_read_at_ms,
            archived: record.archived,
            parent_session_id: record.parent_session_id.clone(),
            parent_run_id: record.parent_run_id.clone(),
            parent_agent_id: record.parent_agent_id.clone(),
            summary: record
                .summary
                .as_ref()
                .map(|summary| SessionSummaryResponse {
                    text: summary.text.clone(),
                    through_message_id: summary.through_message_id.clone(),
                    created_at_ms: summary.created_at_ms,
                    source_message_count: summary.source_message_count,
                }),
            context_trimmed: record.context_trimmed.as_ref().map(|trimmed| {
                SessionContextTrimmedResponse {
                    dropped_through_message_id: trimmed.dropped_through_message_id.clone(),
                    at_ms: trimmed.at_ms,
                }
            }),
            message_count: view.message_count,
            preview: view.preview.clone(),
            active_runs: view.active_runs,
            pending_approvals: 0,
            unread: view.unread,
            capabilities: view.capabilities.into(),
            matched: view.matched.as_ref().map(|found| SessionMatchResponse {
                message_id: found.message_id.clone(),
                snippet: found.snippet.clone(),
            }),
        }
    }
}

#[derive(Clone, Debug, Serialize, ToSchema)]
#[serde(rename_all = "camelCase")]
pub(crate) struct SessionsEnvelope {
    pub(crate) sessions: Vec<SessionResponse>,
    pub(crate) next_cursor: Option<String>,
}

impl From<&SessionPage> for SessionsEnvelope {
    fn from(page: &SessionPage) -> Self {
        Self {
            sessions: page.sessions.iter().map(SessionResponse::from).collect(),
            next_cursor: page.next_cursor.clone(),
        }
    }
}

#[derive(Clone, Debug, Serialize, ToSchema)]
pub(crate) struct SessionEnvelope {
    pub(crate) session: SessionResponse,
}

#[derive(Clone, Debug, Serialize, ToSchema)]
pub(crate) struct SessionAttachmentResponse {
    #[serde(rename = "type")]
    pub(crate) attachment_type: String,
    pub(crate) name: String,
}

#[derive(Clone, Debug, Serialize, ToSchema)]
#[serde(rename_all = "camelCase")]
pub(crate) struct SessionMessageResponse {
    pub(crate) id: String,
    pub(crate) role: String,
    pub(crate) text: String,
    /// Metadata only; attachment contents are not repeated here.
    pub(crate) attachments: Vec<SessionAttachmentResponse>,
    /// Only `toolCalls`, `toolCallId`, `stepId`, `runId`, `toolStatus`
    /// (`success` or `error`), `toolDurationMs`, `incomplete`, `stopped`,
    /// `revised`, `steer`, `skill`, `clientRequestId`, `communication`,
    /// `kind`, and `source`; other keys are not exposed.
    pub(crate) metadata: BTreeMap<String, Value>,
    pub(crate) created_at_ms: u64,
    /// Present (true) only on silent check-in messages, with `includeHidden=true`.
    #[serde(skip_serializing_if = "std::ops::Not::not")]
    pub(crate) hidden: bool,
}

impl From<&PageMessage> for SessionMessageResponse {
    fn from(entry: &PageMessage) -> Self {
        let message = &entry.message;
        Self {
            id: message.id.clone(),
            role: match message.role {
                MessageRole::User => "user",
                MessageRole::Assistant => "assistant",
                MessageRole::System => "system",
                MessageRole::Tool => "tool",
            }
            .into(),
            text: message.content.text.clone(),
            attachments: message
                .content
                .attachments
                .iter()
                .flatten()
                .map(|attachment| SessionAttachmentResponse {
                    attachment_type: match attachment.attachment_type {
                        AttachmentType::File => "file",
                        AttachmentType::Image => "image",
                        AttachmentType::Url => "url",
                    }
                    .into(),
                    name: attachment.name.clone(),
                })
                .collect(),
            metadata: message
                .content
                .metadata
                .iter()
                .flatten()
                .filter(|(key, _)| EXPOSED_MESSAGE_METADATA.contains(&key.as_str()))
                .map(|(key, value)| (key.clone(), data_value_to_json(value)))
                .collect(),
            created_at_ms: message.created_at_ms,
            hidden: entry.hidden,
        }
    }
}

#[derive(Clone, Debug, Serialize, ToSchema)]
#[serde(rename_all = "camelCase")]
pub(crate) struct SessionMessagesEnvelope {
    pub(crate) messages: Vec<SessionMessageResponse>,
    pub(crate) next_before: Option<String>,
}

impl From<&MessagePage> for SessionMessagesEnvelope {
    fn from(page: &MessagePage) -> Self {
        Self {
            messages: page
                .messages
                .iter()
                .map(SessionMessageResponse::from)
                .collect(),
            next_before: page.next_before.clone(),
        }
    }
}

/// `POST /api/agents/{id}/sessions`; an empty body means `{}`.
#[derive(Clone, Debug, Default, Deserialize, ToSchema)]
#[serde(deny_unknown_fields)]
pub(crate) struct SessionCreateRequest {
    #[serde(default)]
    pub(crate) title: Option<String>,
}

/// `PATCH /api/agents/{id}/sessions/{sid}`; at least one field.
#[derive(Clone, Debug, Deserialize, ToSchema)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub(crate) struct SessionUpdateRequest {
    #[serde(default)]
    pub(crate) title: Option<String>,
    #[serde(default)]
    pub(crate) archived: Option<bool>,
    #[serde(default)]
    pub(crate) last_read_at_ms: Option<u64>,
}
